plan/logical_plan_builder.go

```diff
@@ -1285,45 +1285,46 @@ func (b *planBuilder) buildUpdate(update *ast.UpdateStmt) LogicalPlan {
 }
 
 func (b *planBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.Assignment, p LogicalPlan) ([]*expression.Assignment, LogicalPlan) {
-	schema := p.Schema()
-	newList := make([]*expression.Assignment, schema.Len())
-	modifyColumns := make(map[string]struct{}, schema.Len()) // which columns are in set list.
+	modifyColumns := make(map[string]int, p.Schema().Len()) // which columns are in set list.
 	for _, assign := range list {
-		col, err := schema.FindColumn(assign.Column)
-		if err != nil {
-			b.err = errors.Trace(err)
-			return nil, nil
-		}
-		if col == nil {
-			b.err = errors.Trace(errors.Errorf("column %s not found", assign.Column.Name.O))
-			return nil, nil
+		col, offset, err := p.Schema().FindColumnAndIndex(assign.Column)
+		if err == nil && col == nil {
+			err = errors.Trace(errors.Errorf("column %s not found", assign.Column.Name.O))
 		}
-		offset := schema.ColumnIndex(col)
-		if offset == -1 {
-			b.err = errors.Trace(errors.Errorf("could not find column %s.%s", col.TblName, col.ColName))
-			return nil, nil
-		}
-		newExpr, np, err := b.rewrite(assign.Expr, p, nil, false)
 		if err != nil {
 			b.err = errors.Trace(err)
 			return nil, nil
 		}
-		p = np
-		newList[offset] = &expression.Assignment{Col: col.Clone().(*expression.Column), Expr: newExpr}
 		columnFullName := fmt.Sprintf("%s.%s.%s", col.DBName.L, col.TblName.L, col.ColName)
-		modifyColumns[columnFullName] = struct{}{}
+		modifyColumns[columnFullName] = offset
 	}
 	// If columnes in set list contains generated columns, raise error.
 	for _, tn := range tableList {
 		tableInfo := tn.TableInfo
 		for _, colInfo := range tableInfo.Columns {
-			columnFullName := fmt.Sprintf("%s.%s.%s", tn.Schema.L, tn.Name.L, colInfo.Name.L)
-			if _, ok := modifyColumns[columnFullName]; ok && len(colInfo.GeneratedExprString) != 0 {
-				b.err = ErrBadGeneratedColumn.GenByArgs(colInfo.Name.O, tableInfo.Name.O)
-				return nil, nil
+			if len(colInfo.GeneratedExprString) != 0 {
+				columnFullName := fmt.Sprintf("%s.%s.%s", tn.Schema.L, tn.Name.L, colInfo.Name.L)
+				if _, ok := modifyColumns[columnFullName]; ok {
+					b.err = ErrBadGeneratedColumn.GenByArgs(colInfo.Name.O, tableInfo.Name.O)
+					return nil, nil
+				}
 			}
 		}
 	}
+
+	newList := make([]*expression.Assignment, 0, p.Schema().Len())
+	for _, assign := range list {
+		col, err := p.Schema().FindColumn(assign.Column)
+		var newExpr expression.Expression
+		var np LogicalPlan
+		newExpr, np, err = b.rewrite(assign.Expr, p, nil, false)
```

Comment:
because in next PR, we should not only rewrite expressions in `assignmentList`, but also those in `virtualGeneratedAssignmentList`.